Mask bare RPMs

lubomir opened the pull request:
This implements the masking @contyk described in #33 when resolving dependencies for packages in modules. However that is not sufficient to fix the problem. It works in the transaction created for the module, but all RPMs that are found in that transaction are added to the pile, and then tested again separately.

At that step even RPMs from a module are handled as if they were bare, because the modular dependencies are no longer visible.

That's what the second commit is trying to fix. It immediately marks the added packages as tested. 

With both patches, the reproducer from #33 is working correctly. Testing on real data it seems to remove a lot of packages though, so there is probably a reason why it was there in the first place.

lubomir committed:
Mask bare RPMs when a module provides the same RPM

If RPM with the same name is present in available module and as a bare
RPM, the modular one should have precedence. This applies both when
installing modules or pulling in a bare RPM (which has default streams
of modules available).

Fixes: #33

lubomir committed:
Do not test new entries in the pile

When a new modular RPM gets added to the pile, if we test it again, it
will be tested only against default modules and bare RPMs. The correct
module dependencies will not be used.

Care is needed to not mark modules themselves as tested, since that
could cause some alternative dependency paths to not be processed.

igor-raits commented on main.c:
```diff
@@ -777,6 +779,42 @@ resolve_all_solvables (Pool  *pool,
+                          Id p = q.elements[j];
+                          Solvable *candidate = pool_id2solvable (pool, p);
+                          gboolean is_modular = FALSE;
+                          for (Id *i = available_modular_packages; *i; i++) {
```
there should be function `queue_contains` in fus

lubomir replied:
Good point, but this is iterating over array of `Id`s returned by `pool_whatprovides_ptr`, not a `Queue`.

lubomir commented:
I updated the PR to apply the same masking when resolving dependencies of a bare RPM.

lubomir commented:
Testing the first commit by itself indicates that no packages go missing, so omitting the extra depsolving seems to not cause any problems.

lubomir commented:
Modules have to be explicitly checked to make sure all dependency combinations are explored. The last version of the patch does that.

lubomir commented:
Current status:
* When working on a bare RPM or a module, any modular package will mask any bare RPM with the same name.
* If an RPM is added in a transaction as a dependency of something, it is immediately marked as tested and not examined again.

Running my test compose the patch seems to do the correct thing. It excludes a few packages, but all changes look correct.

r4f4 commented on main.c:
```diff
@@ -694,12 +695,60 @@ _install_transaction (Pool         *pool,
+          gboolean is_modular = FALSE;
+          for (Id *i = available_modular_packages; *i; i++)
+            if (*i == p)
+              is_modular = TRUE;
```
You could break from the loop here.

lubomir replied:
Good suggestion, I'll do that.

r4f4 commented on main.c:
```diff
@@ -694,12 +695,60 @@ _install_transaction (Pool         *pool,
+      queue_init (&q);
+      selection_solvables (pool, &sel, &q);
+
+      for (int j = 0; j < q.count; j++)
```
I wonder if you could replace this for loop with `selection_subtract` to remove the modular packages from the selection.

igor-raits replied:
agree

lubomir committed:
Mark packages from disabled modules as not considered

If the module is disabled, we don't want to mask any bare/default RPMs
just because there is a package in that disabled module.

There are two changes here: the masking function ignores modular
packages that are not considered, and when we disable unavailable
module, we also disable all packages in it.

lubomir commented:
I found an issue with the code: it did not take into account disabled modules. The latest commit fixes that. I have also removed the nested loop and replaced it with `queue_contains`.

igor-raits commented on main.c:
```diff
@@ -693,12 +694,103 @@ _install_transaction (Pool         *pool,
+  g_auto(Queue) available_modular_pkgs;
+  queue_init(&available_modular_pkgs);
+  for (Id *pp = modular_packages; *pp; pp++)
+    {
```
no need for `{}`

igor-raits commented on main.c:
```diff
@@ -693,12 +694,103 @@ _install_transaction (Pool         *pool,
+  for (Id *pp = modular_packages; *pp; pp++)
+    {
+      if (map_tst (pool->considered, *pp))
+        {
```
no need for `{}`

igor-raits commented on main.c:
```diff
@@ -693,12 +694,103 @@ _install_transaction (Pool         *pool,
+      queue_init (&sel);
+      selection_make (pool, &sel, name, SELECTION_NAME);
+      if (!sel.count)
+        {
```
no need for `{}`

igor-raits commented on main.c:
```diff
@@ -693,12 +694,103 @@ _install_transaction (Pool         *pool,
+        {
+          Id p = q.elements[j];
+          if (!queue_contains (&available_modular_pkgs, p))
+            {
```
no need for `{}`

igor-raits commented on main.c:
```diff
@@ -693,12 +694,103 @@ _install_transaction (Pool         *pool,
+ * @pool: initialized pool
+ * @module: Id of the module to be disabled
+ *
+ * Set the module and all packages in it as not considered. The packages would
```
I don't like removal of considering packages. I think we should just disable module solvable. Can you rescribe case which would be impossible to do if packages from module are not disabled?

lubomir replied:
It would be possible to not disable the packages, but it would mean in `mask_bare_rpms` the way of selecting which modular packages are visible would have to change. Basically what's needed there is *list of all modular packages than are in available modules*. Current implementation asks for everything that provides *modular-package()* and then removes packages that are not considered. How can I get the list of packages better?

igor-raits replied:
hmmm, there is some function `pool_is_trivially_installable` or smth like that... it might do the trick.

I will look at it when I have time

igor-raits commented on main.c:
```diff
@@ -740,8 +832,9 @@ resolve_all_solvables (Pool  *pool,
           if (!g_str_has_prefix (pool_id2str (pool, s->name), "module:"))
             {
               g_debug ("Installing %s:", pool_solvid2str (pool, p));
+              mask_bare_rpms(pool);
```
style: `func ()`

igor-raits commented on main.c:
```diff
@@ -773,8 +866,14 @@ resolve_all_solvables (Pool  *pool,
                   for (; *pp; pp++)
-                    if (!queue_contains (&t, *pp))
-                      map_clr (pool->considered, *pp);
+                    {
```
please stop adding those useless braces ;)

lubomir replied:
They are not useless, they make it much easier to add debugging output during development :)

igor-raits commented on main.c:
```diff
@@ -693,12 +694,103 @@ _install_transaction (Pool         *pool,
+  for (int k = 0; k < q.count; k++)
+    {
+      Id p = q.elements[k];
+      //g_debug ("  - %s", pool_solvid2str (pool, p));
```
why is it commented?

igor-raits commented on main.c:
```diff
@@ -693,12 +694,103 @@ _install_transaction (Pool         *pool,
+static void
+disable_module (Pool *pool, Id module)
+{
+  // g_debug ("Disabling %s", pool_solvid2str(pool, module));
```
why is it commented?

igor-raits commented on main.c:
```diff
@@ -693,12 +694,103 @@ _install_transaction (Pool         *pool,
+    {
+      Id pp = available_modular_pkgs.elements[i];
+      Solvable *modpkg = pool_id2solvable (pool, pp);
+      const gchar *name = pool_id2str (pool, modpkg->name);
```
no need t ocreate this variable

igor-raits commented on main.c:
```diff
@@ -693,12 +694,103 @@ _install_transaction (Pool         *pool,
+    {
+      if (map_tst (pool->considered, *pp))
+        {
+          queue_push2 (&available_modular_pkgs, SOLVER_SOLVABLE | SOLVER_INSTALL, *pp);
```
no, this is wrong. Do selection, then substract unneeded solvables. don't create some overengineered stuff

lubomir committed:
Fix style issues

lubomir committed:
Disable non-default modules when solving bare RPMs

Otherwise a package in non-default stream could block a bare RPM that
should actually be available.

lubomir commented:
I rebased this on latest master.
